lib/runtime/life/resolver.go

```diff
@@ -933,11 +933,11 @@ func ext_crypto_ed25519_verify_version_1(vm *exec.VirtualMachine) int64 {
 	}
 
 	if sigVerifier.IsStarted() {
-		signature := runtime.Signature{
-			PubKey:    pubKey.Encode(),
-			Sign:      signature,
-			Msg:       message,
-			KeyTypeID: crypto.Ed25519Type,
+		signature := crypto.Signing{
+			PubKey:     pubKey.Encode(),
+			Sign:       signature,
+			Msg:        message,
+			VerifyFunc: ed25519.VerifySignature,
```

Comment:
The caller here should wrap the error with the signature type used. Since it's not aware of the signature type after this point, I suggest you use a wrapping helper function such as

```go
func verifySignatureWrapper(signatureType string, verifyFunc SigVerifyFunc) SigVerifyFunc {
	return func(publicKey, signature, message []byte) (err error) {
		err = verifyFunc(publicKey, signature, message)
		if err != nil {
			return fmt.Errorf("%s: %w", signatureType, err)
		}
		return nil
	}
}
```

which you can then use with

```suggestion
			VerifyFunc: verifySignatureWrapper("ed25519", ed25519.VerifySignature),
```

This applies to other fields in this file.

Reply:
I don't see the necessity of using a wrapper here, each signature already returns it's sig type when an error occurs, for example, `ed25519`:

```go
pubKey, err := NewPublicKey(publicKey)
if err != nil {
     return err
}

_, err = pubKey.Verify(message, signature)
if err != nil {
     return fmt.Errorf("ed25519: %w: for message 0x%x, signature 0x%x and public key 0x%x",
	    crypto.ErrSignatureVerificationFailed, message, signature, publicKey)
}
```

In my mind this could be resolved just this, instead of the above:

```go
pubKey, err := NewPublicKey(publicKey)
if err != nil {
     return fmt.Errorf("ed25519: %w", err)
}
...
```

Reply:
I don't see the necessity for using the wrapper here, since each signature already adds it's sign type when an error occurs, maybe i'm not seeing something

Reply:
Ok thinking about it, my suggestion makes it just stranger, so never mind!

But to explain myself, the `VerifySignature` of i.e. `ed25519` should not be responsible for wrapping the error with its signature type ideally, it should only be concerned on which parts of itself failed. You would also have to wrap every error returned inside VerifySignature with the prefix `ed25519: `.

For example I prefer

```go
func divide(a, b int) (result int, err error) {
  if b == 0 {
     return fmt.Errorf("cannot divide by zero")
  } else if b < 0 {
     return fmt.Errof("cannot divide with a negative number")
  }
  return a / b
}

func caller() error {
  result, err := divide(5, 2)
  if err != nil {
    return fmt.Errorf("divide: %w", err)
  }
}
```

compared to

```go
func divide(a, b int) (result int, err error) {
  if b == 0 {
     return fmt.Errorf("divide: cannot divide by zero")
  } else if b < 0 {
     return fmt.Errof("divide: cannot divide with a negative number")
  }
  return a / b
}

func caller() error {
  result, err := divide(5, 2)
  if err != nil {
    return err
  }
}
```

Reply:
ohhh now I get it, at first I thought that was something wrong with the code, but it was a suggestion, my bad haha 😅

Reply:
No worry! Your code is good, it's just me being a bit of an error wrapping extremist :smile_cat: